tests/test_spyre_cb.py

```diff
@@ -0,0 +1,282 @@
+        executor_class = Executor.get_class(vllm_config)
+
+        engine_core = EngineCore(vllm_config=vllm_config,
+                                 executor_class=executor_class,
+                                 log_stats=True)
+        # engine = V1LLMEngine.from_engine_args(engine_args)
+
+        tokenizer = AutoTokenizer.from_pretrained(model)
+        prompt_tokens1 = tokenizer(prompt1).input_ids
+        prompt_tokens2 = tokenizer(prompt2).input_ids
+
+        # # add first request
+        engine_core.add_request(
+            EngineCoreRequest(
+                request_id=uuid.uuid4(),
+                prompt=prompt1,
+                prompt_token_ids=prompt_tokens1,
+                mm_inputs=None,
+                mm_hashes=None,
+                mm_placeholders=None,
+                sampling_params=sampling_params1,
+                eos_token_id=None,
+                arrival_time=time.time(),
+                lora_request=None,
+            ))
+        assert len(engine_core.scheduler.waiting) == 1
+        assert len(engine_core.scheduler.running) == 0
+
+        _ = engine_core.step()
+        assert len(engine_core.scheduler.waiting) == 0
+        assert len(engine_core.scheduler.running) == 1
+
+        engine_core.add_request(
+            EngineCoreRequest(
+                request_id=uuid.uuid4(),
+                prompt=prompt2,
+                prompt_token_ids=prompt_tokens2,
+                mm_inputs=None,
+                mm_hashes=None,
+                mm_placeholders=None,
+                sampling_params=sampling_params2,
+                eos_token_id=None,
+                arrival_time=time.time(),
+                lora_request=None,
+            ))
+        assert len(engine_core.scheduler.waiting) == 1
+        assert len(engine_core.scheduler.running) == 1
+
+        _ = engine_core.step()
+        assert len(engine_core.scheduler.waiting) == 0
+        assert len(engine_core.scheduler.running) == 1
+
+        _ = engine_core.step()
+        assert len(engine_core.scheduler.waiting) == 0
+        assert len(engine_core.scheduler.running) == 2
+
+        _ = engine_core.step()
+        assert len(engine_core.scheduler.waiting) == 0
+        assert len(engine_core.scheduler.running) == 2
+
+        _ = engine_core.step()
+        assert len(engine_core.scheduler.waiting) == 0
+        assert len(engine_core.scheduler.running) == 2
+
+        _ = engine_core.step()
+        assert len(engine_core.scheduler.waiting) == 0
+        assert len(engine_core.scheduler.running) == 2
+
+        _ = engine_core.step()
+        assert len(engine_core.scheduler.waiting) == 0
+        assert len(engine_core.scheduler.running) == 2
+
+        _ = engine_core.step()
+        assert len(engine_core.scheduler.waiting) == 0
+        assert len(engine_core.scheduler.running) == 2
+
+        _ = engine_core.step()
+        assert len(engine_core.scheduler.waiting) == 0
+        assert len(engine_core.scheduler.running) == 2
```

Comment:
Hmm, it seems like this goes on for more than I understand, request[0] should have hit the `stop_reason` already and ejected from `running` queue, but it stays here until it hits its `max_tokens` 🤔 Perhaps I'm understanding this wrong.

Reply:
" Stop strings require detokenizing the output. Detokenization is handled in the output processor that runs at the AsyncLLM and LLMEngine level
So, we stream tokens from EngineCore -> AsyncLlm. AsyncLLM checks for stop strings in the OutputProcessor class. If a stop string is detected, we stop generating and send an abort message back to EngineCore to terminate the request in the EngineCore " 

This is why the request wasn't stopping using the `stop` string